File: backend/src/models/medicine.model.js
const { DataTypes } = require("sequelize");
const { sequelize } = require("../config/db");

const Medicine = sequelize.define(
  "Medicine",
  {
    id: {
      type: DataTypes.BIGINT.UNSIGNED,
      autoIncrement: true,
      primaryKey: true,
    },
    userId: {
      type: DataTypes.BIGINT.UNSIGNED,
      allowNull: false,
      references: {
        model: "users",
        key: "id",
      },
    },
    name: {
      type: DataTypes.STRING(150),
      allowNull: false,
      validate: {
        notEmpty: { msg: "Medicine name is required." },
        len: { args: [1, 150], msg: "Name must be between 1 and 150 characters." },
      },
    },
    dosage: {
      type: DataTypes.STRING(50),
      allowNull: false,
      validate: {
        notEmpty: { msg: "Dosage is required." },
      },
    },
    form: {
      type: DataTypes.ENUM("tablet", "capsule", "syrup", "injection", "drops", "inhaler", "other"),
      allowNull: false,
    },
    frequency: {
      type: DataTypes.STRING(50),
      allowNull: false,
    },
    times: {
      type: DataTypes.JSON,
      allowNull: false,
      defaultValue: [],
      validate: {
        isValidTimes(value) {
          if (!Array.isArray(value) || value.length === 0) {
            throw new Error("At least one reminder time is required.");
          }
        },
      },
    },
    startDate: {
      type: DataTypes.DATEONLY,
      allowNull: false,
    },
    endDate: {
      type: DataTypes.DATEONLY,
      allowNull: true,
    },
    instructions: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    color: {
      type: DataTypes.STRING(7),
      allowNull: false,
      defaultValue: "#0EA5A0",
      validate: {
        is: {
          args: /^#[0-9A-Fa-f]{6}$/,
          msg: "Color must be a valid hex code (e.g., #0EA5A0).",
        },
      },
    },
    stockCount: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 30,
      validate: {
        min: { args: [0], msg: "Stock count cannot be negative." },
      },
    },
    lowStockThreshold: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 5,
      validate: {
        min: { args: [0], msg: "Threshold cannot be negative." },
      },
    },
  },
  {
    tableName: "medicines",
    timestamps: true,
    createdAt: "createdAt",
    updatedAt: "updatedAt",
  }
);

module.exports = Medicine;

File: frontend/lib/models/medicine_model.dart
import 'package:flutter/material.dart';

enum MedicineForm { tablet, capsule, syrup, injection, drops, inhaler, other }

extension MedicineFormX on MedicineForm {
  String get label {
    switch (this) {
      case MedicineForm.tablet:
        return 'Tablet';
      case MedicineForm.capsule:
        return 'Capsule';
      case MedicineForm.syrup:
        return 'Syrup';
      case MedicineForm.injection:
        return 'Injection';
      case MedicineForm.drops:
        return 'Drops';
      case MedicineForm.inhaler:
        return 'Inhaler';
      case MedicineForm.other:
        return 'Other';
    }
  }

  IconData get icon {
    switch (this) {
      case MedicineForm.tablet:
        return Icons.medication_outlined;
      case MedicineForm.capsule:
        return Icons.medication_liquid_outlined;
      case MedicineForm.syrup:
        return Icons.local_drink_outlined;
      case MedicineForm.injection:
        return Icons.vaccines_outlined;
      case MedicineForm.drops:
        return Icons.water_drop_outlined;
      case MedicineForm.inhaler:
        return Icons.air_outlined;
      case MedicineForm.other:
        return Icons.healing_outlined;
    }
  }
}

class Medicine {
  final String id;
  String name;
  String dosage; // e.g. "500 mg"
  MedicineForm form;
  String frequency; // e.g. "Twice a day"
  List<TimeOfDay> times;
  DateTime startDate;
  DateTime? endDate;
  String instructions; // e.g. "After food"
  Color color;
  int stockCount;
  int lowStockThreshold;

  Medicine({
    required this.id,
    required this.name,
    required this.dosage,
    required this.form,
    required this.frequency,
    required this.times,
    required this.startDate,
    this.endDate,
    this.instructions = '',
    required this.color,
    this.stockCount = 30,
    this.lowStockThreshold = 5,
  });

  bool get isLowStock => stockCount <= lowStockThreshold;
}
